pypeit/images/rawimage.py

```diff
@@ -19,399 +19,701 @@
+        #     when adding the shot noise.
+        self.rn2img = self.build_rn2img(units='ADU')
+        self.var = self.rn2img.copy()
+
+        #   - Subtract the overscan.  Uncertainty from the overscan subtraction
+        #     is added to the variance.
+        if self.par['use_overscan']:
             self.subtract_overscan()
-        if par['trim']:
+
+        # TODO: Do we need to keep trim and orient as optional?
+        #   - Trim to the data region.  This trims the image, rn2img, var, and
+        #     datasec_img.
+        if self.par['trim']:
             self.trim()
-        if par['orient']:
+
+        #   - Re-orient to PypeIt convention. This re-orients the image, rn2img,
+        #     var, and datasec_img.
+        if self.par['orient']:
             self.orient()
-        if par['use_biasimage']:  # Bias frame, if it exists, is *not* trimmed nor oriented
+            
+        #   - Check the shape of the bpm
+        if self.bpm.shape != self.image.shape:
+            msgs.error(f'Bad-pixel mask has incorrect shape: found {self.bpm.shape}, expected '
+                       f'{self.image.shape}.  The shape must match a trimmed and oriented '
+                       'PypeIt-processed image.')
+            
+        #   - Subtract master bias
+        if self.par['use_biasimage']:
+            # Bias frame.  Shape and orientation must match *processed* image,
+            # and the units *must* be ADU.  Uncertainty from the bias
+            # subtraction is added to the variance.
             self.subtract_bias(bias)
-        if par['use_darkimage']:  # Dark frame, if it exists, is TODO:: check: trimmed, oriented (and oscan/bias subtracted?)
-            self.subtract_dark(dark)
-        if par['apply_gain']:
+            
+        #   - Convert from ADU to electrons/counts.  This converts the image,
+        #     the variance, and the detector variance (because the latter is
+        #     used by `add_shot_noise`).  TODO: I don't think there's any
+        #     particular reason why this could not be done early, but it's
+        #     important that it happen before we subtract the dark so that the
+        #     dark can have a different exposure time.
+        if self.par['apply_gain']:
             self.apply_gain()
+            
+        #   - Subtract master dark.
+        if self.par['use_darkimage']:
+            # Dark frame.  Shape and orientation must match *processed* image,
+            # and the units *must* be in electrons/counts.  Uncertainty from the
+            # dark subtraction is added to the variance.
+            # TODO: With exptime now included in PypeItImage, we can explicitly
+            # check and/or account for the difference in exposure time between
+            # the dark image and the image being processed.
+            self.subtract_dark(dark)
+        
+        # Use the bias-subtracted and dark-subtracted counts to include shot
```

Comment:
I would add oscan subtracted to this list.

Reply:
This as all changed now.